core/src/main/java/io/undertow/server/protocol/framed/ShutdownFallbackExecutor.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2014 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.undertow.server.protocol.framed;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link ShutdownFallbackExecutor} wrapper around a single threaded executor
+ * which will execute pending tasks when the worker has been shut down.
+ */
+final class ShutdownFallbackExecutor {
+    private static volatile Executor EXECUTOR = null;
+    private ShutdownFallbackExecutor() {
+        // Static Utility
+    }
+
+    static void execute(Runnable runnable) {
```

Comment:
I think this idea is safer, however this executor needs to shut down when it has run out of tasks. Setting the core pool size to zero does not really accomplish this, as the JDK executor does not work the way you would expect it to (it won't grow past core threads until the queue is full, so in this case I don't think this executor will actually function, I don't know if it has special case coding for a core pool size of zero however).

Reply:
This works the same as `Executors.newCachedThreadPool()` with a max size of one rather than `Integer.MAX_VALUE`, and a much shorter keepalive time (100ms down from 1 minute).

Shutting down the executor would require additional synchronization that I'd prefer to avoid if possible.

threads will all be reaped 100ms (though we might as well drop that down to 10ms here) after the last task they execute. Lazily creating the executor is a (likely unnecessary) optimization, we could statically define it.